### ISSUE TITLE
Allow the transport suggest an application scheduler

### ISSUE THREAD

**Issue by tmds** [aspnet/KestrelHttpServer#1662]
Request to support a no dispatch mode with the libuv transport.

cfr https://github.com/aspnet/KestrelHttpServer/issues/1390#issuecomment-293384383

**Comment by tmds** [comment]
https://github.com/aspnet/KestrelHttpServer/pull/1695 implements part of this.
The other part is to no longer make the ListenerPrimary handle requests but only do dispatching. This is to avoid accepting new connections gets blocked by request handling.

**Comment by davidfowl** [comment]
Kestrel.Core has the mode exposed on its options. If the mode is set to default then the transport gets to pick the scheduler. This lets the transport pick where application callbacks run. By default, our transports will dispatch application code to the thread pool when appropriate. For example, on linux, the sockets transports already dispatches so there's no need to do it again. On windows however, callbacks can run on the IO thread, so it would dispatch by default.

If the user always wants to dispatch despite what the transports pick for the application scheduler, then it can override by setting the SchedulingMode to TheadPool. If the user wants to run on whatever thread they happen to be on basically disabling all dispatching then they can choose Inline.

```C#
public interface IConnectionInformation
{
    IScheduler ApplicationScheduler { get; }
}

public class KestrelServerOptions
{
    public SchedulingMode ApplicationSchedulingMode { get; set; }
}

public enum SchedulingMode
{
    // Use the transport provided ApplicationScheduler 
    Default,
    // Use the thread pool
    ThreadPool,
    // Run on the "current" thread
    Inline
}
```

One issue is that using the `IScheduler` type doesn't have an overload that accepts a state object. This means we'd need to either change that or we decide to move `IThreadPool` to Transport.Abstractions. I'd rather change the `IScheduler` interface.

/cc @pakrym

**Comment by davidfowl** [comment]
Filed https://github.com/dotnet/corefxlab/issues/1495

**Comment by tmds** [comment]
This is specifically for Libuv. When dispatching inline, it is safer to no longer handle requests on the ListenerPrimary. This is to avoid blocking the accept Thread. [`base.DispatchConnection`](https://github.com/aspnet/KestrelHttpServer/blob/42d82a507da9c7522ae64c1a4c5cdd6556035162/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv/Internal/ListenerPrimary.cs#L101) should not called.

**Comment by muratg** [comment]
See https://github.com/aspnet/KestrelHttpServer/issues/1750

**Comment by muratg** [comment]
@davidfowl I'm moving this out. If you think you can do it in the next week or so, please bring it back.

**Comment by muratg** [comment]
For 2.1. the plan is to finish the work.

**Comment by muratg** [comment]
@davidfowl And further...

**Comment by davidfowl** [comment]
😄

**Comment by davidfowl** [comment]
I'm moving this back to preview2 @muratg and @halter73. We should just complete the existing PR you have today and expose something on SocketOptions to control the defaults.

**Comment by davidfowl** [comment]
Punting this from preview2. After running some experiments this change isn't worth it. Though it does clean up the fact that the scheduler is the only abstraction used inside of kestrel, there's no gain to be had at the moment. I'm moving it to the backlog.

https://github.com/aspnet/KestrelHttpServer/pull/2348

**Comment by zpodlovics** [comment]
Would it be possible to have a per core shared nothing and message passing scheduling? [1] [2] It may  possible to support Seastar HTTP as transport [3] [5] for Kestrel. Please note Windows also have official DPDK support [4].

[1] http://www.seastar-project.org/shared-nothing/
[2] http://www.seastar-project.org/message-passing/
[3] http://www.seastar-project.org/http-performance/
[4] https://www.slideshare.net/LF_DPDK/lfdpdk17making-networking-apps-scream-on-windows-with-dpdk
[5] https://github.com/scylladb/seastar/wiki/HTTPD-benchmark

**Comment by davidfowl** [comment]
Closing this in favor of https://github.com/aspnet/KestrelHttpServer/issues/2429
